openssl/src/stack.rs

```diff
@@ -0,0 +1,275 @@
+use std::marker;
+use std::ops;
+use std::mem;
+use std::iter;
+
+use libc::c_void;
+
+use ffi;
+use opaque::Opaque;
+
+/// Trait implemented by stackable types. This must *only* be
+/// implemented on opaque types that can be directly casted into their
+/// `CType`.
+pub trait Stackable {
+    /// Underlying C type for this object.
+    type CType;
+    /// C stack type for this element. Generally called
+    /// `stack_st_{ELEMENT_TYPE}`, normally hidden by the
+    /// `STACK_OF(ELEMENT_TYPE)` macro in the OpenSSL API.
+    type StackType;
+    /// When the stack is owned this destructor is used to free each
+    /// element.
+    fn destructor() -> unsafe extern fn(*mut Self::CType);
```

Comment:
I think we can avoid having to add this by manually popping and dropping all values in the destructor.